src/Nethermind/Nethermind.Network/P2P/Subprotocols/Les/StatusMessageSerializer.cs

```diff
@@ -0,0 +1,291 @@
+            if (message.BufferLimit.HasValue)
+            {
+                bufferLimitLength = Rlp.LengthOf(StatusMessage.KeyNames.BufferLimit) + Rlp.LengthOf(message.BufferLimit.Value);
+                totalContentLength += Rlp.LengthOfSequence(bufferLimitLength);
+            }
+
+            int maxRechargeRateLength = 0;
+            if (message.MaximumRechargeRate.HasValue)
+            {
+                maxRechargeRateLength = Rlp.LengthOf(StatusMessage.KeyNames.MaximumRechargeRate) + Rlp.LengthOf(message.MaximumRechargeRate.Value);
+                totalContentLength += Rlp.LengthOfSequence(maxRechargeRateLength);
+            }
+
+            int maxRequestCostsLength = 0;
+            int innerCostListLength = 0;
+            if (message.MaximumRequestCosts != null)
+            {
+                // todo - what's the best way to do this? Calculating the length twice is definitely less than ideal.
+                // Maybe build RLP for them here, and append bytes below?
+                maxRequestCostsLength += Rlp.LengthOf(StatusMessage.KeyNames.MaximumRequestCosts);
+                foreach (var item in message.MaximumRequestCosts)
+                {
+                    int costContentLength = Rlp.LengthOf(item.MessageCode) + Rlp.LengthOf(item.BaseCost) + Rlp.LengthOf(item.RequestCost);
+                    innerCostListLength += Rlp.LengthOfSequence(costContentLength);
+                }
+                maxRequestCostsLength += Rlp.LengthOfSequence(innerCostListLength);
+                totalContentLength += Rlp.LengthOfSequence(maxRequestCostsLength);
+            }
+            #endregion
+
+            #region Encode Values
+            int totalLength = Rlp.LengthOfSequence(totalContentLength);
+            byteBuffer.EnsureWritable(totalLength);
+            rlpStream.StartSequence(totalContentLength);
+
+            rlpStream.StartSequence(protocolVersionLength);
+            rlpStream.Encode(StatusMessage.KeyNames.ProtocolVersion);
+            rlpStream.Encode(message.ProtocolVersion);
+
+            rlpStream.StartSequence(networkIdLength);
+            rlpStream.Encode(StatusMessage.KeyNames.ChainId);
+            rlpStream.Encode(message.ChainId);
+
+            rlpStream.StartSequence(headTdLength);
+            rlpStream.Encode(StatusMessage.KeyNames.TotalDifficulty);
+            rlpStream.Encode(message.TotalDifficulty);
+
+            rlpStream.StartSequence(headHashLength);
+            rlpStream.Encode(StatusMessage.KeyNames.BestHash);
+            rlpStream.Encode(message.BestHash);
+
+            rlpStream.StartSequence(headNumLength);
+            rlpStream.Encode(StatusMessage.KeyNames.HeadBlockNo);
+            rlpStream.Encode(message.HeadBlockNo);
+
+            rlpStream.StartSequence(genesisHashLength);
+            rlpStream.Encode(StatusMessage.KeyNames.GenesisHash);
+            rlpStream.Encode(message.GenesisHash);
+
+            if (message.AnnounceType.HasValue)
```

Comment:
if it does not have value then you probably need to encode empty sequence here - cannot just skip it

Reply:
ok, it looks like LES has some very strange way of encoding